lib/langchain/llm/azure.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+module Langchain::LLM
+  # LLM interface for Azure OpenAI Service APIs: https://learn.microsoft.com/en-us/azure/ai-services/openai/
+  #
+  # Gem requirements:
+  #    gem "ruby-openai", "~> 4.0.0"
+  #
+  # Usage:
+  #    openai = Langchain::LLM::Azure.new(api_key:, llm_options: {}, embedding_deployment_url: chat_deployment_url:)
```

Comment:
Since it's just the last part of the URL that changes specifying the `model_name`, e.g.:
```
https://custom-domain.openai.azure.com/openai/deployments/gpt-35-turbo
https://custom-domain.openai.azure.com/openai/deployments/ada-2
```

Any reason why we wouldn't treat them like model names and adhere to the interface that all other LLMs follow, like such:
```ruby
    DEFAULTS = {
      chat_completion_model_name: "gpt-35-turbo",
      embeddings_model_name: "ada-2"
      ...
    }.freeze
```

And then, when initializing the client, we'd just append the model name:
```ruby
uri_base: (AZURE_URL + model_name)
```

The idea is that there are defaults but you can override them by specifying the model names you'd like to use.

Reply:
I like this idea and I think it would work but I wonder if it might be too limiting. The last part of the url is actually the `deployment_name`, when you create your deployment on Azure you pick a model and then create a deployment name. I used the model names in the examples and we could force that convention.

However it is possible to have multiple Azure deployments with the same model, like for example I'm not sure if there is a reason for this but someone might want to have separate development and production deployments with names like `gpt-3-5-development` and `gpt-3-5-production`. 

They also allow you to customize each model and train it on your own dataset or fine tune it, though that does end up resulting in a new model with a specific name (from the [API docs](https://learn.microsoft.com/en-us/azure/ai-services/openai/tutorials/fine-tune?tabs=command-line#deploy-fine-tuned-model)): 

````
fine_tuned_model 

Retrieve this value from your fine-tuning job results in the previous step. 
It will look like gpt-35-turbo-0613.ft-b044a9d3cf9c4228b5d393567f693b83. 
You will need to add that value to the deploy_data json
````

It also appears that they have models with the same model name but different versions, for example `text-embedding-ada-002` is the same model name for both version 1 and 2 which [are different](https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models#embeddings). See screenshot below from my Azure dashboard model options:

<img width="1222" alt="Screen Shot 2023-11-02 at 10 33 28 AM" src="https://github.com/andreibondarev/langchainrb/assets/1707425/b842c8ff-8797-4577-a741-ccba667bb390">

Reply:
@duncantmiller Thank you for this detailed explanation, and I agree with your original approach now!

Reply:
I was just playing around with my Azure account a bit and realized I can create multiple services with different custom domain names (the first part of the url) so that could solve the development / production separation. So for example you could have:

`https://foo-development.openai.azure.com/openai/deployments/gpt-35-turbo`

and

`https://foo-production.openai.azure.com/openai/deployments/gpt-35-turbo`